### PR TITLE
[MPC] Specify circuit representation

### PR DESCRIPTION
**Pull request by jschneider-bensch**
This PR provides a circuit representation based on SINE's [`garble_lang`](https://github.com/sine-fdn/garble-lang) project, but heavily simplified.

### CODE REVIEW

File: mpc-spec/src/circuit.rs [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+    InvalidGate(usize),
+    /// The specified output gate does not exist in the circuit.
+    InvalidOutputWire(usize),
+    /// The circuit does not specify any output gates.
+    EmptyOutputSpecification,
+    /// The circuit does not specify input wires.
+    EmptyInputSpecification,
+    /// The circuit specifies a zero-width input.
+    InvalidInputSpecification,
+}
```
Comment (keks) [inline mpc-spec/src/circuit.rs:127]: Maybe also implement `core::fmt::Display` for the error enum?

**Review by keks — CHANGES_REQUESTED** [review]
I think this looks really good! The feedback is mostly nitpicks to make iteration more consistent.

File: mpc-spec/src/circuit.rs [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+            }
+            CircuitError::EmptyInputSpecification => {
+                write!(f, "Circuit does not specify any party inputs.")
+            }
+            CircuitError::InvalidInputSpecification => {
+                write!(f, "Circuit specifies an empty party input.")
+            }
+        }
+    }
+}
```
Comment (keks) [inline mpc-spec/src/circuit.rs:161]: I am not sure if we have a clear policy here, but the [documentation of the standard library Error trait](https://doc.rust-lang.org/stable/std/error/trait.Error.html) says

> Errors must describe themselves through the [Display](https://doc.rust-lang.org/stable/std/fmt/trait.Display.html) and [Debug](https://doc.rust-lang.org/stable/std/fmt/trait.Debug.html) traits. Error messages are typically concise lowercase sentences without trailing punctuation

Personally I think it makes sense to follow that convention, but like I said, not sure if we have a policy for this.

File: mpc-spec/src/circuit.rs [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+
+/// Specifies how many input bits a party is expected to contribute to the
+/// evaluation.
+pub type InputWidth = usize;
+/// Representation of a circuit evaluated by an MPC engine.
+#[derive(Debug, Clone)]
+pub struct Circuit {
+    /// The bit-width of the inputs expected by the different parties,
+    /// [InputWidth] at index `i` representing the number of input bits for
```
Comment (keks) [inline mpc-spec/src/circuit.rs:98]:
```suggestion
    /// [`InputWidth`] at index `i` representing the number of input bits for
```

File: mpc-spec/src/circuit.rs [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+//!
+
+/// Data type to uniquely identify gate output wires.
+pub type WireIndex = usize;
+
+/// An input gate or a logic gate with its input wire specification.
+#[derive(Debug, Clone)]
+pub enum WiredGate {
+    /// An input wire, with its value coming directly from one of the parties.
+    /// Its [WireIndex] must refer to its own gate index.
```
Comment (keks) [inline mpc-spec/src/circuit.rs:75]:
```suggestion
    /// Its [`WireIndex`] must refer to its own gate index.
```

File: mpc-spec/src/circuit.rs [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+    Input(WireIndex),
+    /// A logical XOR gate attached to the two specified input wires. The
+    /// [WireIndex] of each input wire must refer to a lower index than the
```
Comment (keks) [inline mpc-spec/src/circuit.rs:78]:
```suggestion
    /// [`WireIndex`] of each input wire must refer to a lower index than the
```

File: mpc-spec/src/circuit.rs [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+    /// gate's own index.
+    Xor(WireIndex, WireIndex),
+    /// A logical AND gate attached to the two specified input wires. The
+    /// [WireIndex] of each input wire must refer to a lower index than the
```
Comment (keks) [inline mpc-spec/src/circuit.rs:82]:
```suggestion
    /// [`WireIndex`] of each input wire must refer to a lower index than the
```

File: mpc-spec/src/circuit.rs [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+    /// gate's own index.
+    And(WireIndex, WireIndex),
+    /// A logical NOT gate attached to the specified input wire. The
+    /// [WireIndex] of the input wire must refer to a lower index than the
```
Comment (keks) [inline mpc-spec/src/circuit.rs:86]:
```suggestion
    /// [`WireIndex`] of the input wire must refer to a lower index than the
```

**Comment by keks** [comment]
Just a few more nits. Do you know if we have a code style guide for us or so that covers error messages? Otherwise I'd go with the recommendation from the standard library.

**Review by keks — COMMENTED** [review]
Just a few more nits.

Do you know if there is a code style guide for this project that says something about error messages? Otherwise I think it would be best to follow the convention laid out in the standard library.

File: mpc-spec/src/circuit.rs (hunk 2 above)
Comment (jschneider-bensch) [reply mpc-spec/src/circuit.rs:161]: AFAIK we don't have our own policy here, so it makes sense to follow the standard library approach. Thanks for pointing this out!

**Review by keks — APPROVED** [review]
This looks great!

**Pull request merged**